nnstreamer_example/example_cam/nnstreamer_example_cam.c

```diff
@@ -190,7 +190,7 @@ main (int argc, char **argv)
   /** init pipeline */
   str_pipeline =
       g_strdup_printf
-      ("v4l2src name=cam_src ! videoscale ! "
+      ("v4l2src name=cam_src ! videoconvert ! "
```

Comment:
As we all know, we have used  a 'videoconvert' element instead of 'videoscale' in the `nnstreamer_example_filter.c` file. 
@jaeyun-jung  PTAL, Please check and confirm this change.

Reply:
Hmm... do you mean the vivid driver? When it simulates a webcam (sudo modprobe vivid input_types=0x0 node_types=0x01), it indeed defaults to the non-standard 640x360 resolution, so a scaler is required. 
So, if you want to be able to support both vivid and real cameras,  it is a matter of adding both videoscale and videoconvert to the pipeline.
I just pushed such change to my repository.

Reply:
I strongly advice against using v4l2loopback. There are several reasons why we never merged this upstream. You can't really replace a media Kernel driver by an userspace component. One of the reasons is that userspace can be swapped. Some parts of the media core subsystems will cause dead locks if swap happens, because of the type of semaphores that are used to serialize some stuff there. So, any system tainted with v4l2looback driver becomes unstable/unreliable.

So, I strongly suggest to always use either the vivid driver or a real camera.

Reply:
@mchehab, Thank you for your comment. The below example is how to run a producer and a consumer model using v4l2loopback. Could you tell us a gstreamer example to run producer and consumer model using the vivid kernel module that you commented? 

```
## gstreamer: producer
$ sudo modprobe v4l2loopback
$ gst-launch-1.0 videotestsrc ! v4l2sink device=/dev/video0

## gstreamer: consumer 
$ gst-launch-1.0 v4l2src device=/dev/video0 ! xvimagesink 
$ gst-launch-1.0 v4l2src device=/dev/video0 ! videoconvert ! ximagesink (for VNC env)

```

Reply:
Just to be sure that we're talking about the same thing, I suspect that what you call "producer" is called "V4L2 output device" at V4L2 terminology. Using gstreamer terminology, it is basically a sink device.
 A device like a camera ("consumer" on your terminology) is called "V4L2 capture device". It is basically a gstreamer source.

vivid is not a loopback driver. While it may create multiple V4L2 devices, each device is either a capture or an output V4L2 device. You can see its documentation at:
    https://linuxtv.org/downloads/v4l-dvb-apis-new/v4l-drivers/vivid.html
You can control the image it generates via standard V4L2 controls:
   https://imagebin.ca/v/4UPFoaZQx3aO
On this example, I'm showing with qv4l2, with is part of v4l-utils - unfortunately, Tizen package doesn't generate the utilities - just the libraries, but you should be able to install it on Debian/Ubuntu. You can also control it via v4l2-ctl command line application, also part of v4l-utils .deb package.

The default format it generates is similar to the one produced by gstreamer videotestsrc.
So, with vivid,  a pipeline like:

$ gst-launch-1.0 v4l2src device=/dev/video0 ! videoconvert ! xvimagesink 

will produce about the same result as the above pipelines. 

Now, if you really want, instead, to have a device that uses the same device node for capture and playback, you could use, instead, vim2m driver. This is a test driver that does a few image transformations inside the Kernel (software based), like horizontal/vertical image flip. It supports 2 formats, both 16 bits per pixel: RGB565X and YUYV. It may also output the same image as the original input, just like v4l2loopback.

That's said, I just did a quick test on it, using default gstreamer-1.0 plugins as found on Ubuntu 18.04, and it seems that gstream drivers don't properly detect it. It is probably mis-interpreting the m2m flags. I already pinged Gst v4l2 plugin maintainers in order to check what's going wrong with Gstreamer detection of vim2m device nodes.